### PR TITLE
feat: cm secret patch dry run

### PR DESCRIPTION
**Pull request by RohitRaj011**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes https://github.com/devtron-labs/sprint-tasks/issues/1136

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Review by AbhishekA1509 — CHANGES_REQUESTED** [review]

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretContainer.tsx [hunk 1]
```diff
@@ -360,24 +375,45 @@ export const ConfigMapSecretContainer = ({
...
+            } else {
+                updateSearchParams({ headerTab: ConfigHeaderTabType.VALUES })
+            }
         }
-    }, [cmSecretStateLabel, draftData])
+    }, [isLoading, cmSecretStateLabel, draftData])
+
+    // METHOD FOR DEFAULT MERGE STRATEGY
+    const setDefaultMergeStrategy = () => {
+        const _configMapSecretData = draftData ? JSON.parse(draftData.data).configData[0] : configMapSecretData
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretContainer.tsx:405]: not required now

File: src/Pages/Shared/ConfigMapSecret/types.ts [hunk 2]
```diff
@@ -105,6 +111,7 @@ export interface ConfigMapSecretUseFormProps {
     esoSecretYaml: string
     hasCurrentDataErr: boolean
     isResolvedData: boolean
+    mergeStrategy: ConfigToolbarProps['mergeStrategy']
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Shared/ConfigMapSecret/types.ts:114]: not required now

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/SelectMergeStrategy.tsx [hunk 3]
```diff
@@ -27,13 +28,20 @@ const SelectMergeStrategy = ({
     handleMergeStrategyChange = noop,
     isDisabled = false,
     variant = 'dropdown',
+    hidePatchOption,
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/SelectMergeStrategy.tsx:31]: this is already being handled in ConfigToolbar

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretDryRun.tsx [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+            cmSecretStateLabel === CM_SECRET_STATE.INHERITED
+                ? inheritedConfigMapSecretData
+                : publishedConfigMapSecretData
+
+        if (draftData) {
+            if (draftData.action === DraftAction.Delete) {
+                configMapSecretData = inheritedConfigMapSecretData
+            } else if (draftData.draftState === DraftState.Init || draftData.draftState === DraftState.AwaitApproval) {
+                configMapSecretData = {
+                    ...JSON.parse(draftData.data).configData?.[0],
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretDryRun.tsx:83]: not required now

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretDryRun.tsx [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+    // CONSTANTS
+    const isCreateView = id === null
+    const shouldPrompt = isCreateView && isFormDirty
+
+    // PROMPT FOR UNSAVED CHANGES
+    usePrompt({ shouldPrompt })
+
+    const dryRunConfigMapSecretData = useMemo(() => {
+        let configMapSecretData: CMSecretConfigData =
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretDryRun.tsx:73]: early returns not possible
moved to util

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretApproveButton.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+} from '@devtron-labs/devtron-fe-common-lib'
+
+import { hasApproverAccess, importComponentFromFELibrary } from '@Components/common'
+
+import { hasHashiOrAWS } from './utils'
+import { ConfigMapSecretApproveButtonProps } from './types'
+
+const ApproveRequestTippy = importComponentFromFELibrary('ApproveRequestTippy', null, 'function')
+
+export const ConfigMapSecretApproveButton = ({
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretApproveButton.tsx:16]: some utils are present only in dashboard, these are being used at other places as well

[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretApproveButton.tsx:16]: Its only two small utils which should ideally be in common 😅 , but can leave if taking time

File: src/Pages/Shared/ConfigMapSecret/utils.ts [hunk 7]
```diff
@@ -697,6 +760,67 @@ export const getConfigMapSecretResolvedData = (
...
+}
+
 export const getConfigMapSecretError = <T extends unknown>(res: PromiseSettledResult<T>) =>
     res.status === 'rejected' && res.reason?.code !== ERROR_STATUS_CODE.NOT_FOUND ? res.reason : null
+
+export const parseConfigMapSecretSearchParams = (searchParams: URLSearchParams): ConfigMapSecretQueryParamsType => {
+    const headerTab = searchParams.get('headerTab') as ConfigMapSecretQueryParamsType['headerTab']
+
+    return {
+        headerTab: Object.values(ConfigHeaderTabType).includes(headerTab) ? headerTab : null,
```
Comment (AbhishekA1509) [inline src/Pages/Shared/ConfigMapSecret/utils.ts:823]: Can we use CONFIG_HEADER_TAB_VALUES

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**